rct_optimizations/include/rct_optimizations/types.h

```diff
@@ -44,46 +49,80 @@ struct Pose6d
+    , in_target(Eigen::Matrix<double, WORLD_DIM, 1>::Zero())
   {
   }
 
-  Correspondence2D3D(const Eigen::Vector3d& in_target_,
-                     const Eigen::Vector2d& in_image_)
+  Correspondence(const Eigen::Matrix<double, IMAGE_DIM, 1>& in_image_,
+                 const Eigen::Matrix<double, WORLD_DIM, 1>& in_target_)
     : in_target(in_target_)
     , in_image(in_image_)
   {
   }
 
-  Eigen::Vector3d in_target;
-  Eigen::Vector2d in_image;
-};
-using CorrespondenceSet = std::vector<Correspondence2D3D>;
+  /** @brief N-dimensional location of the feature relative to the sensor */
+  Eigen::Matrix<double, IMAGE_DIM, 1> in_image;
 
-struct Correspondence3D3D
+  /** @brief N-dimensional location of the feature relative to the target origin */
+  Eigen::Matrix<double, WORLD_DIM, 1> in_target;
+};
+/** @brief Typedef for correspondence between 2D feature in image coordinates and 3D feature in target coordinates */
+using Correspondence2D3D = Correspondence<2, 3>;
+/** @brief Typedef for correspondence between 3D feature in sensor coordinates and 3D feature in target coordinates */
+using Correspondence3D3D = Correspondence<3, 3>;
+
+// Deprecated typedefs
+using CorrespondenceSet [[deprecated]] = Correspondence2D3D::Set;
+using Correspondence3DSet [[deprecated]] = Correspondence3D3D::Set;
+
+/**
+ * @brief A set of data representing a single observation of a calibration target.
+ * This consists of the feature correspondences as well as the transforms to the "mount" frames of the camera and target.
+ * For a moving camera or target, the "mount" pose would likely be the transform from the robot base to the robot tool flange.
+ * For a stationary camera or target, this "mount" pose would simply be identity.
+ *
+ * Note that @ref to_camera_mount and @ref to_target_mount do not necessarily need to be relative the the same coordinate system because
+ * certain calibration problems might optimize a 6D transform in between the root frame of @ref to_camera mount and the root frame of @ref to_target_mount
+ *
+ * Keep in mind that the optimization itself determines the final calibrated transforms from these "mount" frames to the camera and target.
+ */
+template<Eigen::Index IMAGE_DIM, Eigen::Index WORLD_DIM>
+struct Observation
 {
-  Correspondence3D3D()
-    : in_target(Eigen::Vector3d::Zero())
-    , in_image(Eigen::Vector3d::Zero())
+  using Set = std::vector<Observation<IMAGE_DIM, WORLD_DIM>>;
+  Observation()
+    : to_camera_mount(Eigen::Isometry3d::Identity())
+    , to_target_mount(Eigen::Isometry3d::Identity())
   {
   }
 
-  Correspondence3D3D(const Eigen::Vector3d& in_target_,
-                     const Eigen::Vector3d& in_image_)
-    : in_target(in_target_)
-    , in_image(in_image_)
+  Observation(const Eigen::Isometry3d& to_camera_mount_,
+              const Eigen::Isometry3d& to_target_mount_)
+    : to_camera_mount(to_camera_mount_)
+    , to_target_mount(to_target_mount_)
```

Comment:
Would be preferable to name these to describe both the parent and child frames

Reply:
Are you thinking something like `camera_base_to_camera_mount`?

Naming is a little tricky because different optimizations will interpret these transforms differently. For example the hand-eye calibration we make the assumption that the `to_target_mount` and `to_camera_mount_` are relative to the same frame. However in the case where both the target and camera are actuated by kinematic chains we might develop a new calibration algorithm where these transforms are not relative to the same frame and we optimize a transform that connects them (sidebar: maybe we should actually set up the hand-eye calibration like this, and add the ability to choose whether or not that connecting transform is part of the optimization). I initially left the names a little bit vague to represent this